crates/circuit/src/dag_circuit.rs

```diff
@@ -5263,6 +5257,123 @@ impl DAGCircuit {
         Ok(new_node)
     }
 
+    /// Apply a [PackedOperation] to the back of the circuit.
+    #[allow(clippy::too_many_arguments)]
+    pub fn apply_operation_back(
+        &mut self,
+        py: Python,
+        op: PackedOperation,
+        qargs: &[Qubit],
+        cargs: &[Clbit],
+        params: Option<SmallVec<[Param; 3]>>,
+        extra_attrs: Option<Box<ExtraInstructionAttributes>>,
+        #[cfg(feature = "cache_pygates")] py_op: Option<PyObject>,
+        check: bool,
+    ) -> PyResult<NodeIndex> {
+        self.inner_apply_op(
+            py,
+            op,
+            qargs,
+            cargs,
+            params,
+            extra_attrs,
+            #[cfg(feature = "cache_pygates")]
+            py_op,
+            check,
+            false,
+        )
+    }
+
+    /// Apply a [PackedOperation] to the front of the circuit.
+    #[allow(clippy::too_many_arguments)]
+    pub fn apply_operation_front(
+        &mut self,
+        py: Python,
+        op: PackedOperation,
+        qargs: &[Qubit],
+        cargs: &[Clbit],
+        params: Option<SmallVec<[Param; 3]>>,
+        extra_attrs: Option<Box<ExtraInstructionAttributes>>,
+        #[cfg(feature = "cache_pygates")] py_op: Option<PyObject>,
+        check: bool,
+    ) -> PyResult<NodeIndex> {
+        self.inner_apply_op(
+            py,
+            op,
+            qargs,
+            cargs,
+            params,
+            extra_attrs,
+            #[cfg(feature = "cache_pygates")]
+            py_op,
+            check,
+            true,
+        )
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn inner_apply_op(
+        &mut self,
+        py: Python,
+        op: PackedOperation,
+        qargs: &[Qubit],
+        cargs: &[Clbit],
+        params: Option<SmallVec<[Param; 3]>>,
+        extra_attrs: Option<Box<ExtraInstructionAttributes>>,
+        #[cfg(feature = "cache_pygates")] py_op: Option<PyObject>,
+        check: bool,
+        front: bool,
+    ) -> PyResult<NodeIndex> {
+        // Check that all qargs are within an acceptable range
+        qargs.iter().try_for_each(|qarg| {
+            if !(0..self.num_qubits() as u32).contains(&qarg.0) {
+                return Err(PyValueError::new_err(format!(
+                    "Qubit index {} is out of range. This DAGCircuit currently has only {} qubits.",
+                    qarg.0,
+                    self.num_qubits()
+                )));
+            }
+            Ok(())
+        })?;
```

Comment:
As written, these checks iterate over all the bits for every arg, which is quadratic.

You should be able to get away with just converting each arg to a `usize` and making sure it is less than `num_{qu,cl}bits()`.